Build failure on GHC 9.4 on Windows due to linker issues

Attempting to build the `unix-time` test suite with GHC 9.4.2 on Windows fails with the following linker error:

<details>

```
$ cabal get unix-time-0.4.7 && cd unix-time-0.4.7/
Unpacking to unix-time-0.4.7\

$ cabal test -w ghc-9.4.2
Resolving dependencies...
Build profile: -w ghc-9.4.2 -O1
In order, the following will be built (use -v for more details):
 - unix-time-0.4.7 (first run)
Configuring unix-time-0.4.7...
configure: loading site script C:/msys64/etc/config.site
checking for gcc... C:\ghcup\ghc\949A7A~1.2\lib\../mingw/bin\clang.exe
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether C:\ghcup\ghc\949A7A~1.2\lib\../mingw/bin\clang.exe accepts -g... yes
checking for C:\ghcup\ghc\949A7A~1.2\lib\../mingw/bin\clang.exe option to accept ISO C89... none needed
checking how to run the C preprocessor... C:\ghcup\ghc\949A7A~1.2\lib\../mingw/bin\clang.exe -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking xlocale.h usability... no
checking xlocale.h presence... no
checking for xlocale.h... no
checking for strptime_l... no
checking for timegm... no
checking for _mkgmtime... no
checking for _get_current_locale... yes
checking for _create_locale... yes
checking for strtol_l... no
checking for strtoll_l... no
checking for _isspace_l... yes
checking for _isupper_l... yes
checking for _isdigit_l... yes
checking for _isblank_l... yes
configure: creating ./config.status
config.status: creating cbits/config.h
Preprocessing library for unix-time-0.4.7..
Building library for unix-time-0.4.7..
[1 of 5] Compiling Data.UnixTime.Types ( C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\Data\UnixTime\Types.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\Data\UnixTime\Types.o )

Data\UnixTime\Types.hsc:3:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
  |
3 | import Control.Applicative ((<$>), (<*>))
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2 of 5] Compiling Data.UnixTime.Sys ( C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\Data\UnixTime\Sys.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\Data\UnixTime\Sys.o )
[3 of 5] Compiling Data.UnixTime.Diff ( Data\UnixTime\Diff.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\Data\UnixTime\Diff.o )

Data\UnixTime\Diff.hs:46:11: warning: [-Winline-rule-shadowing]
    Rule "Integral->UnixDiffTime" may never fire
      because ‘fromIntegral’ might inline first
    Suggested fix:
      Add an INLINE[n] or NOINLINE[n] pragma for ‘fromIntegral’
   |
46 | {-# RULES "Integral->UnixDiffTime" fromIntegral = secondsToUnixDiffTime #-}
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[4 of 5] Compiling Data.UnixTime.Conv ( Data\UnixTime\Conv.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\Data\UnixTime\Conv.o )

Data\UnixTime\Conv.hs:11:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
   |
11 | import Control.Applicative
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
[5 of 5] Compiling Data.UnixTime    ( Data\UnixTime.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\Data\UnixTime.o )

cbits\strptime.c:275:8: error:
     warning: implicit declaration of function 'isblank_l' is invalid in C99 [-Wimplicit-function-declaration]
                                isblank_l((unsigned char)*buf, locale)) {
                                ^
    |
275 |                             isblank_l((unsigned char)*buf, locale)) {
    |        ^

cbits\strptime.c:701:14: error:
     warning: implicit declaration of function 'timegm' is invalid in C99 [-Wimplicit-function-declaration]
                    time_t t = timegm(tm);
                               ^
    |
701 |                 time_t t = timegm(tm);
    |              ^
2 warnings generated.
Preprocessing test suite 'spec' for unix-time-0.4.7..
Building test suite 'spec' for unix-time-0.4.7..
[1 of 2] Compiling UnixTimeSpec     ( test\UnixTimeSpec.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\spec\spec-tmp\UnixTimeSpec.o )
[2 of 2] Compiling Main             ( test\Spec.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.4.2\unix-time-0.4.7\build\spec\spec-tmp\Main.o )
[3 of 3] Linking C:\\Users\\winferno\\Documents\\Hacking\\C\\unix-time-0.4.7\\dist-newstyle\\build\\x86_64-windows\\ghc-9.4.2\\unix-time-0.4.7\\build\\spec\\spec.exe
ld.lld: error: undefined symbol: timegm
>>> referenced by libHSunix-time-0.4.7-inplace.a(strptime.o):(strptime_l)
>>> referenced by libHSunix-time-0.4.7-inplace.a(strptime.o):(strptime)

ld.lld: error: undefined symbol: isblank_l
>>> referenced by libHSunix-time-0.4.7-inplace.a(strptime.o):(_strptime)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ghc-9.4.2.exe: `clang.exe' failed in phase `Linker'. (Exit code: 1)
```
</details>

On the other hand, this problem does not surface when building with GHC 9.2.4 on Windows:

<details>

```
$ cabal get unix-time-0.4.7 && cd unix-time-0.4.7/
Unpacking to unix-time-0.4.7\

$ cabal test -w ghc-9.2.4
Resolving dependencies...
Build profile: -w ghc-9.2.4 -O1
In order, the following will be built (use -v for more details):
 - unix-time-0.4.7 (first run)
Configuring unix-time-0.4.7...
configure: loading site script C:/msys64/etc/config.site
checking for gcc... C:\ghcup\ghc\9291B8~1.4\lib\../mingw/bin\gcc.exe
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether C:\ghcup\ghc\9291B8~1.4\lib\../mingw/bin\gcc.exe accepts -g... yes
checking for C:\ghcup\ghc\9291B8~1.4\lib\../mingw/bin\gcc.exe option to accept ISO C89... none needed
checking how to run the C preprocessor... C:\ghcup\ghc\9291B8~1.4\lib\../mingw/bin\gcc.exe -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking xlocale.h usability... no
checking xlocale.h presence... no
checking for xlocale.h... no
checking for strptime_l... no
checking for timegm... no
checking for _mkgmtime... yes
checking for _get_current_locale... yes
checking for _create_locale... yes
checking for strtol_l... no
checking for strtoll_l... no
checking for _isspace_l... yes
checking for _isupper_l... yes
checking for _isdigit_l... yes
checking for _isblank_l... no
configure: creating ./config.status
config.status: creating cbits/config.h
Preprocessing library for unix-time-0.4.7..
Building library for unix-time-0.4.7..
[1 of 5] Compiling Data.UnixTime.Types ( C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\Data\UnixTime\Types.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\Data\UnixTime\Types.o )

Data\UnixTime\Types.hsc:3:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
  |
3 | import Control.Applicative ((<$>), (<*>))
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2 of 5] Compiling Data.UnixTime.Sys ( C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\Data\UnixTime\Sys.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\Data\UnixTime\Sys.o )
[3 of 5] Compiling Data.UnixTime.Diff ( Data\UnixTime\Diff.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\Data\UnixTime\Diff.o )

Data\UnixTime\Diff.hs:46:11: warning: [-Winline-rule-shadowing]
    Rule "Integral->UnixDiffTime" may never fire
      because ‘fromIntegral’ might inline first
    Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘fromIntegral’
   |
46 | {-# RULES "Integral->UnixDiffTime" fromIntegral = secondsToUnixDiffTime #-}
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[4 of 5] Compiling Data.UnixTime.Conv ( Data\UnixTime\Conv.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\Data\UnixTime\Conv.o )

Data\UnixTime\Conv.hs:11:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
   |
11 | import Control.Applicative
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
[5 of 5] Compiling Data.UnixTime    ( Data\UnixTime.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\Data\UnixTime.o )
Preprocessing test suite 'spec' for unix-time-0.4.7..
Building test suite 'spec' for unix-time-0.4.7..
[1 of 2] Compiling UnixTimeSpec     ( test\UnixTimeSpec.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\spec\spec-tmp\UnixTimeSpec.o )
[2 of 2] Compiling Main             ( test\Spec.hs, C:\\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\build\spec\spec-tmp\Main.o )
Linking C:\\Users\\winferno\\Documents\\Hacking\\C\\unix-time-0.4.7\\dist-newstyle\\build\\x86_64-windows\\ghc-9.2.4\\unix-time-0.4.7\\build\\spec\\spec.exe ...
Running 1 test suites...
Test suite spec: RUNNING...
Test suite spec: PASS
Test suite logged to:
C:\Users\winferno\Documents\Hacking\C\unix-time-0.4.7\dist-newstyle\build\x86_64-windows\ghc-9.2.4\unix-time-0.4.7\test\unix-time-0.4.7-spec.log
1 of 1 test suites (1 of 1 test cases) passed.
```
</details>

I suspect this is because GHC 9.4 [switched its C toolchain from GCC to Clang](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.4?version_id=09bb4e233a62e1154b34e12b230d7b80facfabb8), although I need to do some more investigation to see why this would affect `unix-time`.

Upon further investigation, this is indeed related to the GCC->Clang toolchain switchover in GHC 9.4. More specifically, the GCC toolchain that ships with GHC 9.2 and earlier links against MSVCRT, while the Clang toolchain that ships with GHC 9.4 and later links against UCRT. While these two C runtimes are mostly compatible with each other—for instance, they both use the same header files—they do have some subtle differences which manifest when building `unix-time`.

(If you'd like a **tl;dr** version of this, skip to the end of this comment.)

-----

First, let's start with the `timegm` linker error. `unix-time`'s `configure.ac` checks for `timegm` in the following two lines:

https://github.com/kazu-yamamoto/unix-time/blob/80fd289c7248f92636f85f3dc7365a4d8e2a7eb9/configure.ac#L10-L11

On Windows, the C header files do not define `timegm` or `_mkgmtime`. Surprisingly, however, the GCC toolchain successfully finds `_mkgmtime`:

```
checking for timegm... no
checking for _mkgmtime... yes
```

While the Clang toolchain does _not_ find `_mkgmtime`:

```
checking for timegm... no
checking for _mkgmtime... no
```

What is going on here? It turns out that the fact that this ever worked on GCC was due to a complete accident. `autoconf`'s `AC_CHECK_FUNCS(_mkgmtime)` macro works by generating a simple C test program that looks roughly like this:

```c
int main() {
  return _mkgmtime();
  return 0;
}
```

`autoconf` will then use the configured C compiler to compile this test program and check if it returns an exit code of 0. If that succeeds, then `autoconf` will think that `_mkgmtime` is provided by the C compiler. Let's try compiling this program with Clang to see what happens:

```
$ clang test.c -o test.exe
test.c:2:10: warning: implicit declaration of function '_mkgmtime' is invalid in C99 [-Wimplicit-function-declaration]
  return _mkgmtime();
         ^
1 warning generated.
ld.lld: error: undefined symbol: _mkgmtime
>>> referenced by C:/msys64/tmp/test-605700.o:(main)
clang: error: linker command failed with exit code 1 (use -v to see invocation)

$ echo $?
1
```

Clang emits an `-Wimplicit-function-declaration` warning, since it can't find `_mkgmtime` in any of the included header files. This is only a warning, not a fatal error, however. Clang then attempts to link the program, but since the linker can't find `_mkgmtime`, it throws a fatal `undefined symbol` error and returns exit code 1.

Now let's try the same thing, but with GCC:

```
$ gcc test.c -o test.exe
test.c: In function 'main':
test.c:2:10: warning: implicit declaration of function '_mkgmtime' [-Wimplicit-function-declaration]
    2 |   return _mkgmtime();
      |          ^~~~~~~~~

$ echo $?
0
```

Like before, the compiler emits an `-Wimplicit-function-declaration` warning. Surprisingly, however, the linker _succeeds_ this time and returns exit code 0. How is this possible? By sheer accident, the linking step on GCC succeeds because MSVCRT happens to export a function named `_mkgmtime`.

```
$ /c/Program\ Files\ \(x86\)/Microsoft\ Visual\ Studio/2019/BuildTools/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/dumpbin.exe -exports /c/Windows/System32/msvcrt.dll | grep "_mkgmtime"
        647  286 00063130 _mkgmtime
        648  287 00063140 _mkgmtime32
        649  288 00063450 _mkgmtime64
```

On the other hand, UCRT does _not_ export a function named `_mkgmtime`, so the linking step fails:

```
$ /c/Program\ Files\ \(x86\)/Microsoft\ Visual\ Studio/2019/BuildTools/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/dumpbin.exe -exports /c/Windows/System32/ucrtbase.dll | grep "_mkgmtime"
        663  296 000A8330 _mkgmtime32
        664  297 000A8340 _mkgmtime64
       1183  49E 00062E80 _o__mkgmtime32
       1184  49F 00062EA0 _o__mkgmtime64
```

In short, the way that `configure.ac` checks for `_mkgmtime` is fragile. The C header files actually do define `_mkgmtime` on Windows, but you have to include `time.h` to use it. The `configure.ac` script generates a test program without `time.h`, however. In order to convince `autoconf` to include `time.h` in the test program, we need to use `AC_CHECK_DECLS` instead. That is:

```diff
diff --git a/cbits/config.h.in b/cbits/config.h.in
index 7ec66c5..2a4cf91 100644
--- a/cbits/config.h.in
+++ b/cbits/config.h.in
@@ -61,7 +61,7 @@
 #undef HAVE__ISUPPER_L

 /* Define to 1 if you have the `_mkgmtime' function. */
-#undef HAVE__MKGMTIME
+#undef HAVE_DECL__MKGMTIME

 /* "Is Linux" */
 #undef IS_LINUX
diff --git a/cbits/win_patch.h b/cbits/win_patch.h
index 1b5b022..774612d 100644
--- a/cbits/win_patch.h
+++ b/cbits/win_patch.h
@@ -126,7 +126,7 @@ struct tm *gmtime_r(const time_t *_time_t, struct tm *_tm);

 struct tm *localtime_r(const time_t *_time_t, struct tm *_tm);

-#if HAVE__MKGMTIME
+#if HAVE_DECL__MKGMTIME
 #define timegm _mkgmtime
 #define HAVE_TIMEGM 1
 #endif
diff --git a/configure.ac b/configure.ac
index 0732702..4e78553 100644
--- a/configure.ac
+++ b/configure.ac
@@ -8,7 +8,7 @@ AC_CHECK_HEADERS(xlocale.h)

 AC_CHECK_FUNCS(strptime_l)
 AC_CHECK_FUNCS(timegm)
-AC_CHECK_FUNCS(_mkgmtime)
+AC_CHECK_DECLS([_mkgmtime], [], [], [[#include <time.h>]])
 AC_CHECK_FUNCS(_get_current_locale)
 AC_CHECK_FUNCS(_create_locale)
 AC_CHECK_FUNCS(strtol_l)
```

This makes the use of `_mkgmtime` work on GCC and Clang alike.

-----

Next, let's look at the `_isblank_l` linker error. As far as I can tell, the C headers on Windows do not provide `_isblank_l` at all. GCC is unable to find `_isblank_l` at all when configuring `unix-time`, so it successfully falls back on an [alternative implementation](https://github.com/kazu-yamamoto/unix-time/blob/80fd289c7248f92636f85f3dc7365a4d8e2a7eb9/cbits/win_patch.c#L15-L20). On the other hand, Clang _is_ able to find `_isblank_l`. Once again, this happens because UCRT happens to export a function named `_isblank_l`:

```
$ /c/Program\ Files\ \(x86\)/Microsoft\ Visual\ Studio/2019/BuildTools/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/dumpbin.exe -exports /c/Windows/System32/ucrtbase.dll | grep "_isblank_l"
        386  181 0006AB00 _isblank_l
```

While MSVCRT does not:

```
$ /c/Program\ Files\ \(x86\)/Microsoft\ Visual\ Studio/2019/BuildTools/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/dumpbin.exe -exports /c/Windows/System32/msvcrt.dll | grep "_isblank_l"
# Nothing
```

What should we do about this? One option is to make `autoconf` stricter so that `-Wimplicit-function-declaration` warnings become errors when compiling test programs, which would make `autoconf` believe that Clang does not have `_isblank_l`. I'm unclear if this would have any adverse consequences on the rest of the `configure.ac` script, however, so I'm somewhat hesistant to pick this option.

Another option is to simply link against UCRT's `_isblank_l` function. While the C header files shipped with Clang on Windows don't provide `_isblank_l` for whatever reason, the [UCRT documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/isblank-iswblank-isblank-l-iswblank-l?view=msvc-170) _does_ guarantee the existence of `_isblank_l`, so it should be stable to link against it. With that in mind, I propose the following patch to make `_isblank_l` work with Clang:

```diff
diff --git a/cbits/win_patch.h b/cbits/win_patch.h
index 1b5b022..53f9f06 100644
--- a/cbits/win_patch.h
+++ b/cbits/win_patch.h
@@ -116,7 +116,11 @@ long long strtoll_l(const char *nptr, char **endptr, int base, _locale_t locale)
 #define isleap_sum(a, b)    isleap((a) % 400 + (b) % 400)
 #endif /* !defined isleap_sum */

-#if !HAVE__ISBLANK_L
+#if HAVE__ISBLANK_L
+#define isblank_l _isblank_l
+// Needed to avoid -Wimplicit-function-declaration warnings
+int _isblank_l(int c, _locale_t  loc);
+#else
 int isblank_l( int c, _locale_t _loc);
 #endif

```

-----

**tl;dr** The following patch should suffice to fix the build with Clang/GHC 9.4:

<details>

```diff
diff --git a/cbits/config.h.in b/cbits/config.h.in
index 7ec66c5..8b60eb4 100644
--- a/cbits/config.h.in
+++ b/cbits/config.h.in
@@ -1,14 +1,18 @@
 /* cbits/config.h.in.  Generated from configure.ac by autoheader.  */

+/* Define to 1 if you have the declaration of `_mkgmtime', and to 0 if you
+   don't. */
+#undef HAVE_DECL__MKGMTIME
+
 /* Define to 1 if you have the <inttypes.h> header file. */
 #undef HAVE_INTTYPES_H

-/* Define to 1 if you have the <memory.h> header file. */
-#undef HAVE_MEMORY_H
-
 /* Define to 1 if you have the <stdint.h> header file. */
 #undef HAVE_STDINT_H

+/* Define to 1 if you have the <stdio.h> header file. */
+#undef HAVE_STDIO_H
+
 /* Define to 1 if you have the <stdlib.h> header file. */
 #undef HAVE_STDLIB_H

@@ -60,9 +64,6 @@
 /* Define to 1 if you have the `_isupper_l' function. */
 #undef HAVE__ISUPPER_L

-/* Define to 1 if you have the `_mkgmtime' function. */
-#undef HAVE__MKGMTIME
-
 /* "Is Linux" */
 #undef IS_LINUX

@@ -87,5 +88,7 @@
 /* Define to the version of this package. */
 #undef PACKAGE_VERSION

-/* Define to 1 if you have the ANSI C header files. */
+/* Define to 1 if all of the C90 standard headers exist (not just the ones
+   required in a freestanding environment). This macro is provided for
+   backward compatibility; new code need not use it. */
 #undef STDC_HEADERS
diff --git a/cbits/win_patch.h b/cbits/win_patch.h
index 1b5b022..b746b53 100644
--- a/cbits/win_patch.h
+++ b/cbits/win_patch.h
@@ -116,7 +116,11 @@ long long strtoll_l(const char *nptr, char **endptr, int base, _locale_t locale)
 #define isleap_sum(a, b)    isleap((a) % 400 + (b) % 400)
 #endif /* !defined isleap_sum */

-#if !HAVE__ISBLANK_L
+#if HAVE__ISBLANK_L
+#define isblank_l _isblank_l
+// Needed to avoid -Wimplicit-function-declaration warnings
+int _isblank_l(int c, _locale_t  loc);
+#else
 int isblank_l( int c, _locale_t _loc);
 #endif

@@ -126,7 +130,7 @@ struct tm *gmtime_r(const time_t *_time_t, struct tm *_tm);

 struct tm *localtime_r(const time_t *_time_t, struct tm *_tm);

-#if HAVE__MKGMTIME
+#if HAVE_DECL__MKGMTIME
 #define timegm _mkgmtime
 #define HAVE_TIMEGM 1
 #endif
diff --git a/configure.ac b/configure.ac
index 0732702..4e78553 100644
--- a/configure.ac
+++ b/configure.ac
@@ -8,7 +8,7 @@ AC_CHECK_HEADERS(xlocale.h)

 AC_CHECK_FUNCS(strptime_l)
 AC_CHECK_FUNCS(timegm)
-AC_CHECK_FUNCS(_mkgmtime)
+AC_CHECK_DECLS([_mkgmtime], [], [], [[#include <time.h>]])
 AC_CHECK_FUNCS(_get_current_locale)
 AC_CHECK_FUNCS(_create_locale)
 AC_CHECK_FUNCS(strtol_l)
```
</details>

--- a/.gitignore
+++ b/.gitignore
@@ -8,3 +8,4 @@ cabal.sandbox.config
 .ghc.environment.*
 # stack
 .stack-work
+autom4te.cache/*

--- a/cbits/config.h.in
+++ b/cbits/config.h.in
@@ -1,5 +1,9 @@
 /* cbits/config.h.in.  Generated from configure.ac by autoheader.  */
 
+/* Define to 1 if you have the declaration of `_mkgmtime', and to 0 if you
+   don't. */
+#undef HAVE_DECL__MKGMTIME
+
 /* Define to 1 if you have the <inttypes.h> header file. */
 #undef HAVE_INTTYPES_H
 
@@ -59,9 +63,6 @@
 
 /* Define to 1 if you have the `_isupper_l' function. */
 #undef HAVE__ISUPPER_L
-
-/* Define to 1 if you have the `_mkgmtime' function. */
-#undef HAVE__MKGMTIME
 
 /* "Is Linux" */
 #undef IS_LINUX

--- a/cbits/win_patch.h
+++ b/cbits/win_patch.h
@@ -116,7 +116,11 @@ long long strtoll_l(const char *nptr, char **endptr, int base, _locale_t locale)
 #define isleap_sum(a, b)    isleap((a) % 400 + (b) % 400)
 #endif /* !defined isleap_sum */
 
-#if !HAVE__ISBLANK_L
+#if HAVE__ISBLANK_L
+#define isblank_l _isblank_l
+// Needed to avoid -Wimplicit-function-declaration warnings
+int _isblank_l(int c, _locale_t  loc);
+#else
 int isblank_l( int c, _locale_t _loc);
 #endif
 
@@ -126,7 +130,7 @@ struct tm *gmtime_r(const time_t *_time_t, struct tm *_tm);
 
 struct tm *localtime_r(const time_t *_time_t, struct tm *_tm);
 
-#if HAVE__MKGMTIME
+#if HAVE_DECL__MKGMTIME
 #define timegm _mkgmtime
 #define HAVE_TIMEGM 1
 #endif

--- a/configure
+++ b/configure
@@ -651,6 +651,7 @@ infodir
 docdir
 oldincludedir
 includedir
+runstatedir
 localstatedir
 sharedstatedir
 sysconfdir
@@ -721,6 +722,7 @@ datadir='${datarootdir}'
 sysconfdir='${prefix}/etc'
 sharedstatedir='${prefix}/com'
 localstatedir='${prefix}/var'
+runstatedir='${localstatedir}/run'
 includedir='${prefix}/include'
 oldincludedir='/usr/include'
 docdir='${datarootdir}/doc/${PACKAGE}'
@@ -973,6 +975,15 @@ do
   | -silent | --silent | --silen | --sile | --sil)
     silent=yes ;;
 
+  -runstatedir | --runstatedir | --runstatedi | --runstated \
+  | --runstate | --runstat | --runsta | --runst | --runs \
+  | --run | --ru | --r)
+    ac_prev=runstatedir ;;
+  -runstatedir=* | --runstatedir=* | --runstatedi=* | --runstated=* \
+  | --runstate=* | --runstat=* | --runsta=* | --runst=* | --runs=* \
+  | --run=* | --ru=* | --r=*)
+    runstatedir=$ac_optarg ;;
+
   -sbindir | --sbindir | --sbindi | --sbind | --sbin | --sbi | --sb)
     ac_prev=sbindir ;;
   -sbindir=* | --sbindir=* | --sbindi=* | --sbind=* | --sbin=* \
@@ -1110,7 +1121,7 @@ fi
 for ac_var in	exec_prefix prefix bindir sbindir libexecdir datarootdir \
 		datadir sysconfdir sharedstatedir localstatedir includedir \
 		oldincludedir docdir infodir htmldir dvidir pdfdir psdir \
-		libdir localedir mandir
+		libdir localedir mandir runstatedir
 do
   eval ac_val=\$$ac_var
   # Remove trailing slashes.
@@ -1263,6 +1274,7 @@ Fine tuning of the installation directories:
   --sysconfdir=DIR        read-only single-machine data [PREFIX/etc]
   --sharedstatedir=DIR    modifiable architecture-independent data [PREFIX/com]
   --localstatedir=DIR     modifiable single-machine data [PREFIX/var]
+  --runstatedir=DIR       modifiable per-process data [LOCALSTATEDIR/run]
   --libdir=DIR            object code libraries [EPREFIX/lib]
   --includedir=DIR        C header files [PREFIX/include]
   --oldincludedir=DIR     C header files for non-gcc [/usr/include]
@@ -1723,6 +1735,52 @@ $as_echo "$ac_res" >&6; }
   eval $as_lineno_stack; ${as_lineno_stack:+:} unset as_lineno
 
 } # ac_fn_c_check_func
+
+# ac_fn_c_check_decl LINENO SYMBOL VAR INCLUDES
+# ---------------------------------------------
+# Tests whether SYMBOL is declared in INCLUDES, setting cache variable VAR
+# accordingly.
+ac_fn_c_check_decl ()
+{
+  as_lineno=${as_lineno-"$1"} as_lineno_stack=as_lineno_stack=$as_lineno_stack
+  as_decl_name=`echo $2|sed 's/ *(.*//'`
+  as_decl_use=`echo $2|sed -e 's/(/((/' -e 's/)/) 0&/' -e 's/,/) 0& (/g'`
+  { $as_echo "$as_me:${as_lineno-$LINENO}: checking whether $as_decl_name is declared" >&5
+$as_echo_n "checking whether $as_decl_name is declared... " >&6; }
+if eval \${$3+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+$4
+int
+main ()
+{
+#ifndef $as_decl_name
+#ifdef __cplusplus
+  (void) $as_decl_use;
+#else
+  (void) $as_decl_name;
+#endif
+#endif
+
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_compile "$LINENO"; then :
+  eval "$3=yes"
+else
+  eval "$3=no"
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
+fi
+eval ac_res=\$$3
+	       { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_res" >&5
+$as_echo "$ac_res" >&6; }
+  eval $as_lineno_stack; ${as_lineno_stack:+:} unset as_lineno
+
+} # ac_fn_c_check_decl
 cat >config.log <<_ACEOF
 This file contains any messages produced by compilers while
 running configure, to aid debugging if configure makes a mistake.
@@ -3304,16 +3362,17 @@ _ACEOF
 fi
 done
 
-for ac_func in _mkgmtime
-do :
-  ac_fn_c_check_func "$LINENO" "_mkgmtime" "ac_cv_func__mkgmtime"
-if test "x$ac_cv_func__mkgmtime" = xyes; then :
-  cat >>confdefs.h <<_ACEOF
-#define HAVE__MKGMTIME 1
-_ACEOF
-
+ac_fn_c_check_decl "$LINENO" "_mkgmtime" "ac_cv_have_decl__mkgmtime" "#include <time.h>
+"
+if test "x$ac_cv_have_decl__mkgmtime" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
 fi
-done
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_DECL__MKGMTIME $ac_have_decl
+_ACEOF
 
 for ac_func in _get_current_locale
 do :

--- a/configure.ac
+++ b/configure.ac
@@ -8,7 +8,7 @@ AC_CHECK_HEADERS(xlocale.h)
 
 AC_CHECK_FUNCS(strptime_l)
 AC_CHECK_FUNCS(timegm)
-AC_CHECK_FUNCS(_mkgmtime)
+AC_CHECK_DECLS([_mkgmtime], [], [], [[#include <time.h>]])
 AC_CHECK_FUNCS(_get_current_locale)
 AC_CHECK_FUNCS(_create_locale)
 AC_CHECK_FUNCS(strtol_l)

